scripts/add_gen3_repo.py

```diff
@@ -0,0 +1,149 @@
+
+
+def main():
+    args = _Parser().parse_args()
+    log = lsst.log.Log.getLogger("add_gen3_repo")
+
+    # To convert consistently, don't use any previous output
+    dataset = ap_verify.dataset.Dataset(args.dataset)
+    gen3_repo = os.path.join(dataset.datasetRoot, "preloaded")
+    if os.path.exists(gen3_repo):
+        log.warn("Clearing out %s and making it from scratch...", gen3_repo)
+        shutil.rmtree(gen3_repo)
+    os.makedirs(gen3_repo)
+
+    mode = "copy" if args.drop_gen2 else "relsymlink"
+
+    log.info("Converting templates...")
+    gen2_templates = dataset.templateLocation
+    _migrate_gen2_to_gen3(dataset, gen2_templates, None, gen3_repo, mode,
+                          config_file="convertRepo_templates.py")
+
+    log.info("Converting calibs...")
+    with tempfile.TemporaryDirectory() as tmp:
+        workspace = ap_verify.workspace.Workspace(tmp)
+        ap_verify.ingestion.ingestDataset(dataset, workspace)
+
+        gen2_repo = workspace.dataRepo
+        gen2_calibs = workspace.calibRepo
+        # Files stored in the Gen 2 part of the dataset, can be safely linked
+        _migrate_gen2_to_gen3(dataset, gen2_repo, gen2_calibs, gen3_repo, mode,
+                              config_file="convertRepo_calibs.py")
+        # Our refcats and defects are temporary files, and must not be linked
+        _migrate_gen2_to_gen3(dataset, gen2_repo, gen2_calibs, gen3_repo, mode="copy",
+                              config_file="convertRepo_copied.py")
+
+    log.info("Exporting Gen 3 registry to configure new repos...")
+    _export_for_copy(dataset, gen3_repo)
+
+
+def _migrate_gen2_to_gen3(dataset, gen2_repo, gen2_calib_repo, gen3_repo, mode, config_file):
+    """Convert a Gen 2 repository into a Gen 3 repository.
+
+    Parameters
+    ----------
+    dataset : `lsst.ap.verify.dataset.Dataset`
+        The dataset being migrated.
+    gen2_repo, gen2_calib_repo : `str`
+       The locations of the original repositories.
+    gen3_repo : `str`
+       The location of the Gen 3 repository. Must exist, but need not be
+       initialized as a repository.
+    mode : {'relsymlink', 'copy'}
+       Whether the Gen 3 repo should contain symbolic links to the Gen 2
+       datasets, or an independent copy.
+    mode : `str`
+       The config file (in the dataset config directory) with a configuration
+       for `~lsst.obs.base.gen2to3.ConvertRepoTask`
+    """
+    instrument = _get_instrument_class(dataset.camera)
+    config = os.path.join(dataset.configLocation, config_file)
+
+    # calib/<instrument>, refcats, and skymaps collections created by default
+    convertGen2RepoToGen3.convert(gen2_repo, gen3_repo, instrument,
+                                  # skymap detected automatically; do nothing
+                                  calibs=gen2_calib_repo,
+                                  config=config,
+                                  transferMode=mode)
```

Comment:
Did you consider using `ConvertRepoTask` directly instead of the command-line-driver wrapper function here?  I don't want to get in the way of you getting things done by asking you to change things unnecessarily, but I'd like the `Task` to be the interface usually used in Python code (as is usual for `Tasks`), so if there's anything the script does that it can't, I'd like to know about it.

(We do have test code that calls the command-line-script wrapper function, but that's just to test as much of the command-line script as we can.)

Reply:
Yes, I did consider it, but then I would have to duplicate about 30 lines of code (https://github.com/lsst/obs_base/blob/master/python/lsst/obs/base/script/convertGen2RepoToGen3.py#L110-L138), manually setting up separate skymap configs (though I don't understand why that's necessary) and a Butler. The code would then need to be kept in sync with any changes to `obs_base`, which is difficult given that an isolated script like this is susceptible to bit rot.

This approach seems more maintainable, because it delegates as much logic as possible to well-maintained code.